Don't flush ImageBufferIOSurfaceBackend on creation.

mattwoodrow opened the pull request:
#### 2f426da28a3bc3d40eae3d98b687b686d714055f
<pre>
Don&apos;t flush ImageBufferIOSurfaceBackend on creation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=259998">https://bugs.webkit.org/show_bug.cgi?id=259998</a>
&lt;rdar://113657395&gt;

Reviewed by Kimmo Kinnunen.

ImageBufferIOSurfaceBackend currently clears the newly allocated IOSurface and then synchronously flushes during construction.
Rather than flushing immediately, we should be able to defer this flush until a consumer attempts to read the contents of the surface.

This adds a m_needsFirstFlush variable to track the omitted initial flush, and ensures we include the flush when the pixels are accessed.

It also set RemoteImageBufferProxy&apos;s m_needsFlush to true, so that remote consumers that use the ImageBufferBackendHandle to access the buffer&apos;s
pixel also know they need to explicitly flush.

This could potentially be a performance regression for remote consumers that attempt to allocate a surface and then use it for compositing as a blank surface,
since it adds the need to wait on the flush.

It should be a performance progression in the majority of cases though, since we usually draw to the surface (and require a flush for that), and will coalesce the
flushes.

* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::create):
(WebCore::ImageBufferIOSurfaceBackend::flushContextDraws):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::create):

Canonical link: <a href="https://commits.webkit.org/266883@main">https://commits.webkit.org/266883@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5121fca1b628b6e8aead2af9b02ab43d5165489

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14964 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15268 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15622 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16741 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/14078 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/15102 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17789 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15369 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16730 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15144 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15619 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12712 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17446 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12899 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13498 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20462 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13977 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13666 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16903 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14223 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/12028 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13506 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3621 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17839 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/14065 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -90,7 +90,6 @@ std::unique_ptr<ImageBufferIOSurfaceBackend> ImageBufferIOSurfaceBackend::create
         return nullptr;
 
     CGContextClearRect(cgContext.get(), FloatRect(FloatPoint::zero(), backendSize));
-    CGContextFlush(cgContext.get());
 
```
Can't we move the `clearRect()` to `ImageBufferIOSurfaceBackend::context()`?

```
GraphicsContext& ImageBufferIOSurfaceBackend::context()
{
    if (!m_context) {
        m_context = makeUnique<GraphicsContextCG>(ensurePlatformContext());
        applyBaseTransform(*m_context);
        m_context->clearRect(backendSize());
    }
    return *m_context;
}
```

shallawa commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -130,10 +129,10 @@ void ImageBufferIOSurfaceBackend::flushContext()
-        return false;
-    if (!m_context->consumeHasDrawn())
+    bool contextNeedsFlush = m_context && m_context->consumeHasDrawn();
+    if (!contextNeedsFlush && !m_needsFlush)
```
If the `clearRect()` is moved to `ImageBufferIOSurfaceBackend::context()` then we do not need to change this function or to add `m_needsFlush`.

mattwoodrow replied:
Is it possible for a consumer to never access the context though?

I thought it'd be possible to have a caller just use `putPixelBuffer` (or `getPixelBuffer`), and we'd want the buffer to be cleared despite the `GraphicsContextCG` not yet existing.

shallawa replied:
You are right.

But having already two variables about the dirtiness of the IOSurface 

```
ImageBufferIOSurfaceBackend::m_mayHaveOutstandingBackingStoreReferences
GraphicsContextCG::consumeHasDrawn()
```

makes me nervous for adding a third one. Can we avoid adding it? At least it should be renamed to `m_needsFirstFlush`, `m_needsFlushAfterCreation` or something else.

mattwoodrow replied:
I can't think of any way to avoid the extra variable unfortunately, but I've renamed it to `m_needsFirstFlush` to make the usage clearer.

kkinnunen-apple commented on Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
```diff
@@ -130,10 +129,10 @@ void ImageBufferIOSurfaceBackend::flushContext()
-    if (!m_context)
-        return false;
-    if (!m_context->consumeHasDrawn())
+    bool contextNeedsFlush = m_context && m_context->consumeHasDrawn();
```
This seems ok, while unfortunate...

Conceptually these are the same:
-  "GraphicsContextCG has drawn to the surface" -- `consumeHasDrawn()`
 - "CGContext has drawn to the surface" -- `std::exchange(m_needsFirstFlush, false)`

kkinnunen-apple reviewed (commented):
this is a load-bearing flush, intentionally added. 
I think you need to change initialization in RemoteImageBufferProxy::m_needsFlush { true };

As a meta-comment you just cannot remove it with the rationale "We shouldn't need to flush on creation". The commit message should explain the rationale why you think it's correct to remove.

As Simon said, remember to run tests with 
* default
* --no-use-remote-layer-tree --experimental-feature UseGPUProcessForDOMRenderingEnabled=false
* -1

Directories are: fast/canvas http/wpt/2dcontext

mattwoodrow committed:
Don't flush ImageBufferIOSurfaceBackend on creation.
https://bugs.webkit.org/show_bug.cgi?id=259998
<rdar://113657395>

Reviewed by Kimmo Kinnunen.

ImageBufferIOSurfaceBackend currently clears the newly allocated IOSurface and then synchronously flushes during construction.
Rather than flushing immediately, we should be able to defer this flush until a consumer attempts to read the contents of the surface.

This adds a m_needsFirstFlush variable to track the omitted initial flush, and ensures we include the flush when the pixels are accessed.

It also set RemoteImageBufferProxy's m_needsFlush to true, so that remote consumers that use the ImageBufferBackendHandle to access the buffer's
pixel also know they need to explicitly flush.

This could potentially be a performance regression for remote consumers that attempt to allocate a surface and then use it for compositing as a blank surface,
since it adds the need to wait on the flush.

It should be a performance progression in the majority of cases though, since we usually draw to the surface (and require a flush for that), and will coalesce the
flushes.

* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp:
(WebCore::ImageBufferIOSurfaceBackend::create):
(WebCore::ImageBufferIOSurfaceBackend::flushContextDraws):
* Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.h:
* Source/WebKit/WebProcess/GPU/graphics/RemoteImageBufferProxy.h:
* Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBackend.cpp:
(WebKit::ImageBufferShareableMappedIOSurfaceBackend::create):

Canonical link: https://commits.webkit.org/266883@main
